File: src/main/scala/110.balanced-binary-tree.scala
/**
  * Definition for a binary tree node.
  * class TreeNode(var _value: Int) {
  *   var value: Int = _value
  *   var left: TreeNode = null
  *   var right: TreeNode = null
  * }
  */
object Q110 {
  import java.lang.Math
  import common.scala.TreeNode

  def isBalanced(root: TreeNode): Boolean =
    if (root == null) true
    else {
      def maxDepth(root: TreeNode): Int =
        if (root == null) 0
        else {
          val left  = maxDepth(root.left)
          val right = maxDepth(root.right)
          return Math.max(left, right) + 1
        }
      val balance = Math.abs(maxDepth(root.left) - maxDepth(root.right)) <= 1
      balance && isBalanced(root.left) && isBalanced(root.right)
    }

}

File: src/main/java/110.balanced-binary-tree.java
import common.java.TreeNode;

/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode(int x) { val = x; } }
 */
class Q110Java {
  public boolean isBalanced(TreeNode root) {
    if (root == null) {
      return true;
    }
    int left = height(root.left);
    int right = height(root.right);
    return Math.abs(left - right) <= 1 && isBalanced(root.left) && isBalanced(root.right);
  }

  private int height(TreeNode root) {
    if (root == null) {
      return 0;
    }
    return Math.max(height(root.left), height(root.right)) + 1;
  }
}
